### PR TITLE
Rework spinners to track full rotations (part 2)

### LINKED ISSUE #8224: Spinners can be cheesed
**Describe the bug:** 

By tapping extremely rapidly with two fingers with the circle of the spinner at the middle. You can go up to 430 RPM.

**Screenshots or videos showing encountered issue:** 

[Video](https://player.vimeo.com/video/396941387) 

**osu!lazer version:** 

2020.311.0 

**Logs:**
 Android

### PR DESCRIPTION
**Pull request by smoogipoo**
Resolves https://github.com/ppy/osu/issues/8224
Supersedes https://github.com/ppy/osu/pull/24859

In #24859, the player was required to spin a total of 360° in either direction for a full rotation to be counted.

In this PR, the player is required to spin 360° _after starting from 0°_. This is an important distinction, as it means you need to "reset" the spinner to turn it in the other direction.  
As an example, if you first spin 90° clockwise, then change direction, you first "reset" the spinner by spinning 90° counter-clockwise, and then continue in that same direction _another_ 360° to complete a full rotation.

As a result of https://github.com/ppy/osu/commit/9b8628071702a7acae3854eca8e2a3e12a05eb61, this PR also potentially fixes intermittent spinner test failures.

### CODE REVIEW

File: osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        /// </summary>
+        private Turn currentTurn;
+
+        /// <summary>
+        /// Adds an instant of spin.
+        /// </summary>
+        /// <param name="currentTime">The current time.</param>
+        /// <param name="delta">The rate-independent, instantaneous delta of the angle moved through. Negative values represent counter-clockwise movements, positive values represent clockwise movements.</param>
+        /// <returns>The total rotation after applying the delta.</returns>
+        public void AddDelta(double currentTime, float delta)
```
Comment (Walavouchey) [inline osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs:57]: says here it returns something but it doesn't

File: osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+            currentMaxRotation = Math.Max(currentMaxRotation, rotation);
+        }
+
+        /// <summary>
+        /// Removes an instant of spin.
+        /// </summary>
+        /// <param name="currentTime">The current time.</param>
+        /// <param name="delta">The rate-independent, instantaneous delta of the angle moved through. Negative values represent counter-clockwise movements, positive values represent clockwise movements.</param>
+        /// <returns>The total rotation after removing the delta.</returns>
+        public void RemoveDelta(double currentTime, float delta)
```
Comment (Walavouchey) [inline osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs:103]: same as above

File: osu.Game.Rulesets.Osu.Tests/SpinFramesGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+using System;
+using System.Collections.Generic;
+using osu.Game.Rulesets.Osu.Replays;
+using osu.Game.Rulesets.Replays;
+using osuTK;
+
+namespace osu.Game.Rulesets.Osu.Tests
+{
+    public class SpinFramesGenerator
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/SpinFramesGenerator.cs:12]: Because this is a class only used in tests, I would hope that structurally it's very easy to quickly understand.

Due to the self-nesting, right now I find things very hard to follow. Could it be restructured to not be nested?

ie.

```csharp

var generator = new SpinFramesGenerator(double startTime)
generator.Spin(from: 0, to: 60, duration: 1000);
var frames = generator.Build();
```

Or if you're wanting to keep the chaining,

```csharp
var generator = new SpinFramesGenerator(double startTime)
    .Spin(from: 0, to: 60, duration: 1000)
    .Spin(...)
    .Build();
```

should still be possible without the nesting by returning `this`.

File: osu.Game.Rulesets.Osu.Tests/SpinFramesGenerator.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+        private SpinFramesGenerator(SpinFramesGenerator? last, float endAngle, double duration)
+        {
+            this.last = last;
+            startAngle = last?.endAngle ?? endAngle;
+            this.endAngle = endAngle;
+            this.duration = duration;
+        }
+
+        public List<ReplayFrame> Build(double spinnerStartTime, float spinOffset)
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/SpinFramesGenerator.cs:37]: I'm also not immediately sure what `spinOffset` is and how it differs from the `.From()` angle specification. I don't think this is important if the above proposed structural change is applied though.

**Review by peppy — COMMENTED** [review]
Initial pass, still going through the main logic.

File: osu.Game.Rulesets.Osu.Tests/SpinFramesGenerator.cs (hunk 3 above)
Comment (bdach) [reply osu.Game.Rulesets.Osu.Tests/SpinFramesGenerator.cs:12]: Yeah, definitely agree with this. The linked-list-y structure here isn't really doing much here except obfuscating intent. It shouldn't take very much rearranging at all to get rid of it in favour of collecting to like an internal list or something.

File: osu.Game.Rulesets.Osu.Tests/SpinnerSpinHistoryTest.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using NUnit.Framework;
+using osu.Game.Rulesets.Osu.Objects.Drawables;
+
+namespace osu.Game.Rulesets.Osu.Tests
+{
+    [TestFixture]
+    public class SpinnerSpinHistoryTest
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/SpinnerSpinHistoryTest.cs:10]: Hmm, I get that this is intending to have test coverage of various rewind scenarios, but it's quite hard to follow with this level of abstraction. Wondering if we can make `TestSpinnerInput` also cover some rewind cases...

File: osu.Game.Rulesets.Osu.Tests/SpinFramesGenerator.cs (hunk 4 above)
Comment (bdach) [reply osu.Game.Rulesets.Osu.Tests/SpinFramesGenerator.cs:37]: It's positional offset from the centre of the playfield. So with a non-zero `spinOffset`, you aren't spinning around the centre of the playfield exactly, but around some other slightly-perturbed location relative to the centre.

Maybe rename to `spinCentreOffset` or something?

File: osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        private void beginNewTurn(double currentTime, int direction)
+        {
+            turningPoints.Push(currentTurn);
+            currentTurn = new Turn(currentTime, direction) { Angle = currentTurn.Angle };
+        }
+
+        /// <summary>
+        /// Represents a single direction turn of the spinner.
+        /// </summary>
+        private struct Turn
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs:147]: Can we perhaps rename this to ~~"Moment"~~ "TurningPoint" instead of "Turn"?

"One turn" to most people would describe a full rotation, which isn't the case here, so it's quite confusing.

File: osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+            /// </summary>
+            public readonly double StartTime;
+
+            /// <summary>
+            /// The turning direction.
+            /// </summary>
+            public readonly int Direction;
+
+            /// <summary>
+            /// The current angle.
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs:160]:
```suggestion
            /// The current absolute angle.
```
i guess

Comment (peppy) [reply osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs:160]: So to clarify, this "Turn" class is not storing anything but absolute angles right? There's no delta stored here.

Maybe naming it `TurningPoint` is better if that's the case?

**Comment by bdach** [comment]
Not much other to add from me after an initial read-through of the diff that's not there in the existing comment. I will say though that it's a rather difficult one to get through. The rotation accounting I'm still not 100% sure of the correctness of. I'll probably attempt to re-read with a fresher mind when reviews are applied.

File: osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs (hunk 6 above)
Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs:147]: Before I do this, can you please define _all_ terminology here? Because we have different ways of thinking about things - I think of a turn as a varying amount, a spin as a full rotation, and a rotation as an aggregate of all spins. I've already taken a different path from that in this PR which is to think of a "spin" as a "rotation" which my mind already struggles with.

Please define:
- A word for what is basically a arbitrary angle, or what is currently `Turn`.
- A word for the "arbitrary angle" as above, that is currently being updated, or what is currently `currentTurn`.
- A word for the maximum absolute angle in any one direction starting from 0 for the current 360-degree... rotation, or what is currently `currentMaxRotation`.
- A word for the total aggregate rotation - the summation of all full 360-degree spins plus the current partial one (`currentMaxRotation`), or what is currently `TotalRotation`.

Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs:147]: For example, would you feel comfortable with:
```
private struct TurningPoint { ... }

private Stack<TurningPoint> TurningPoints = ...

private TurningPoint currentTurningPoint;

public void AddDelta(...)
{
    currentTurningPoint.Angle = ...
}
```

File: osu.Game.Rulesets.Osu.Tests/SpinFramesGenerator.cs (hunk 4 above)
Comment (smoogipoo) [reply osu.Game.Rulesets.Osu.Tests/SpinFramesGenerator.cs:37]: Removed.

File: osu.Game.Rulesets.Osu.Tests/SpinFramesGenerator.cs (hunk 3 above)
Comment (smoogipoo) [reply osu.Game.Rulesets.Osu.Tests/SpinFramesGenerator.cs:12]: Done.

File: osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs (hunk 6 above)
Comment (peppy) [reply osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs]: I've applied naming changes as discussed on discord, will resolve this for now as it's feeling good to me.

File: osu.Game.Rulesets.Osu.Tests/SpinnerSpinHistoryTest.cs (hunk 5 above)
Comment (smoogipoo) [reply osu.Game.Rulesets.Osu.Tests/SpinnerSpinHistoryTest.cs:10]: I've also added a visual test to that one.

**Comment by bdach** [comment]
The forward logic I'm _somewhat_ confident of the correctness off; the rewind logic - no idea. It's very difficult to follow.

File: osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+
+                // The new segment should be in the same direction and with the excess of the previous segment.
+                currentSpinSegment.CurrentRotation = rotation * direction;
+                currentMaxRotation = 0;
+            }
+
+            currentMaxRotation = Math.Max(currentMaxRotation, rotation);
+        }
+
+        private void rewindDelta(double currentTime, float delta)
```
[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs:104]: That's correct, yes.

**Comment by smoogipoo** [comment]
In the process of trying to fix https://github.com/ppy/osu/pull/24999#discussion_r1350238817, I seem to have lost the plot, and I now feel completely incapable of completing this/coding this for some reason...

The idea I was going for in my latest changes was to use the monotonicity of segments to figure out what the rotation of the segment we arrived at should be, based on the absolute rotation at the current point. It fixes the mentioned case+tests but breaks in the visual rewind test.

Could I ask one of you two to try your hand at implementing the `SpinnerSpinHistory` class from scratch, and seeing how you go? All the tests I have should provide rigorous enough backbone to test the implementation.

**Comment by smoogipoo** [comment]
I think I found the issue with the new rewind impl in the newer test I've added... Using the absolute rotation it's impossible to figure out what the value should be when crossing 0.

The case is:
- Spin 180° CCW.
- Spin 360° CW.
- Rewind to a point where you're in that second spin but in the top-left quadrant.

It's hard to know, using absolute rotation, what the `CurrentRotation` of that second spin should be at that point, given the details we have. In a more graph-y way of thinking about it, there's two possible y-coordinates (rotation values) at that point.
